### DISCUSSION TITLE
Plastic models of soils

### DISCUSSION THREAD

**Discussion by jfenap** [idaholab/moose#20026] [Q&A General] [answered]
Hi all,
At the Geological Institute of Spain, we have develop several plastic models for soils. These models account that the material is made of porous and relate changes in the relative void space (porosity) to plastic dilatancy, representing dilatant and compactive volume changes. For this reason, it is assumed that any change in mean pressure is accompanied by elastic change in volume which makes the bulk modulus non constant. 
The yield is a function with elliptical shape, and after yield the material approaches to a consolidation curve that is determined experimentally. See for example (http://docs.itascacg.com/3dec700/common/models/camclay/doc/modelcamclay.html).
We would like to code these models in MOOSE. We already studied the doco of TensorMechanics, but we are not sure that `ComputeMultipleInelasticStress` is going to work for us.
Where is a good place to start? And is there a similar code on which we can base ours?
Thanks and regards,
Fernando

**Comment by WilkAndy** [comment]
I'm not going to answer your question in this post, sorry, but i wanted to sincerely welcome you to MOOSE.  Your contribution to TensorMechanics will be a truly valuable addition to MOOSE, since i think soil plasticity is one of the important mechanical theories that MOOSE cannot do yet.

I contributed a lot to `ComputeMultipleInelasticStress` in its early days.  Perhaps you can tell us more detail about why it's not going to work for you.  There are lots of people involved in TensorMechanics who might like to suggest potential code designs as this discussion proceeds, such as @bwspenc and @sapitts and @dschwen and @SudiptaBiswas  .  Feel free to look at the github "blame" for relevant classes to identify relevant people, and then mention them here to flag to them that this is a discussion of potential interest.

**Comment by somu15** [comment, marked as answer]
Hi @jfenap 

We have a non-linear soil model called I-Soil in MASTODON, a MOOSE application for structural and soil dynamics. Here is the link to the documentation: https://mooseframework.inl.gov/mastodon/manuals/theory/index.html. In addition, here is the link to the file in MASTODON that implements the I-Soil model: https://github.com/idaholab/mastodon/blob/devel/src/materials/ComputeISoilStress.C. Hopefully, this will be somewhat helpful.

Som
